### ISSUE TITLE
PySide2 Now in Portage – We Did It, Everyone!

### ISSUE THREAD

**Issue by leycec** [waebbl/waebbl-gentoo#237]
O.K., so Gentoo developer @ZeroChaos- [actually did it on our behalf](https://github.com/leycec/raiagent/issues/89), which you should profusely thank him for the next time you see him on the `#gentoo` IRC channel.

Everyone's PySide2-based life has now been dramatically simplified. Going forward, to [paraphrase commentary at the `raiagent` overlay](https://github.com/leycec/raiagent/issues/89):

* The existing PySide2 and shiboken2 ebuilds hosted at `::raiagent` will be preserved for a few weeks (modified to log a deprecation warning that users should strongly prefer official Portage ebuilds) before **these ebuilds are permanently removed.**
* The existing `pyside2-tools` ebuild also hosted at `::raiagent` will be preserved **indefinitely,** as there is no official Portage counterpart at the moment.

To avoid breaking FreeCAD here, the above means that this overlay should probably:

* Remove `raiagent` from the `masters` list in the `metadata/layout.conf` file – unless FreeCAD needs `pyside2-tools`, in which case you're (of course) welcome to continue requiring `raiagent`.
* Remove the following note from the top-level `README.md` file:

> **Note:** To emerge the freecad package, you need the pyside packages from the `::raiagent` overlay at https://github.com/leycec/raiagent.git. Use `eselect repository enable raiagent` before emerging freecad.

Thanks again for all the glorious FreeCAD, @waebbl and friends. Also, buy @ZeroChaos- several Irish whiskies, for he has simplified everyone's life. :coffee:

**Comment by waebbl** [comment]
Wow this is good news! Thanks to you for your great work with the ebuilds and thanks to @ZeroChaos for adding the ebuilds to Gentoo.

For FreeCAD, the `::raiagent` overlay is still needed, as long as the `pyside2-tools` are not in Gentoo either. FreeCAD depends on them for building, so the goal must be to add this package into the Gentoo Portage tree as well. But that's a trivial package, compared to the `shiboken` and `pyside` ebuilds, which I'm sure, will find it's way into the tree too sooner or later.

Thanks again for your effort and good work!

**Comment by fordfrog** [comment]
@leycec @waebbl i just added `dev-python/pyside2-tools-5.15.1` to the main tree :-) thanks Cecil for the ebuild :-) please see commit notes in https://github.com/gentoo/gentoo/commit/4ad700c6777d38c4312473076db545fed88f48cf if interested

@ZeroChaos- i would prefer if you could maintain this package as it is related to `dev-python/pyside2` but if you don't want to then i can maintain it myself. but if you wanna take it, just put yourself there and drop me from there, thanks :-)

if there are some issues with the package, please let me know, preferably file a bug report or provide a patch :-)

**Comment by fordfrog** [comment]
btw, maybe a stupid question, but the tools provided by `qtchooser` are different from the ones provided by `pyside2-tools`, right? i'm asking because of the file collision. `qtchooser` also provides `lupdate` whereas `pyside2-tools` provides `pyside2-lupdate`... anybody able to clarify this?

```
 * Detected file collision(s):
 * 
 *      /usr/bin/uic
 *      /usr/bin/rcc
 *      /usr/bin/designer
 * 
 * Searching all installed packages for file collisions...
 * 
 * Press Ctrl-C to Stop
 * 
 * dev-qt/qtchooser-66:0::gentoo
 *      /usr/bin/designer
 *      /usr/bin/rcc
 *      /usr/bin/uic
 ```

up to that, this is how the install tree looks:

```
# tree /var/tmp/portage/dev-python/pyside2-tools-5.15.1/image
/var/tmp/portage/dev-python/pyside2-tools-5.15.1/image
└── usr
    ├── bin
    │   ├── designer -> qtchooser
    │   ├── pyside2-lupdate
    │   ├── rcc -> qtchooser
    │   └── uic -> qtchooser
    └── share
        ├── doc
        │   └── pyside2-tools-5.15.1
        │       └── README.md.bz2
        └── man
            └── man1
                └── pyside2-lupdate.1.bz2

7 directories, 6 files
```

@waebbl maybe `pyside2-tools` is not needed at all and `qtchooser` utils should be used by `freecad` instead?

**Comment by a17r** [comment]
Ultimately, build systems *must* *not* depend on `qtchooser`, it is for end users only. And it might as well call Qt4 binaries by default, depending on configuration.

Instead, build systems shall be pointed at the Qt5 binaries residing in `/usr/lib64/qt5/bin`, for that we have `qmake-utils.eclass`.

**Comment by waebbl** [comment]
I think the `pyside2-lupdate` is still needed, as the standard Qt5 `lupdate` can't yet extract documentation from python sources, according to the `--help` output of `lupdate`. The `rcc` and `uic` binaries are no longer needed, as new versions of those tools have the `-g python` flag and can produce python code from the `.ui` und `.rc` files.

**Comment by fordfrog** [comment]
@waebbl yes, that makes sense (now). will you do the tests as you wrote in gentoo bugzilla or should i drop the tools part?

**Comment by waebbl** [comment]
@fordfrog Will test the freecad build tomorrow, but this does only speak for freecad and not for packages which are more targeted to python or pure python projects using pyside2. But I think you can safely remove the `tools` USE flag, and we should do this to avoid confusion and collisions.

**Comment by fordfrog** [comment]
@waebbl i have one more question wrt `pyside2-tools`. currently it supports three versions of python, but in the end one installs over the other... i can either change the ebuild to support and install only single python target or i can install `pyside2-lupdate` to different paths so they do not overwrite each other. idk which one to choose. any idea/preference?

**Comment by waebbl** [comment]
@fordfrog I think the latter would be more sensible, as shiboken2 and pyside2 both use PYTHON_TARGETS and can support multiple python implementations.

**Comment by fordfrog** [comment]
@waebbl i dropped `tools` from `pyside2-tools` but did not solve the issue with installation of multiple versions yet as idk whether this package will be needed or not, depending on the results of your tests on the other package.

**Comment by fordfrog** [comment]
@waebbl in the end i decided to finish the `pyside2-tools` ebuild, so here is the fix for the installation: https://github.com/gentoo/gentoo/commit/78bbbabb4d8ff9e2909636c1cfd50aedf3795c3c

**Comment by waebbl** [comment]
@fordfrog Thanks, I put it as a backup in my overlay.
@leycec Thanks for all your good work on those packages. I removed your overlay as a master from my overlay, so you can delete the pyside related packages from your overlay, if you wish to do so. Keep up your great work!

Close this.

**Comment by leycec** [comment]
**Wowzers.** Thanks a dynamically allocated heap to @waebbl, @fordfrog, @ZeroChaos-, and everyone else whose GitHub usernames I have conveniently forgotten for resurrecting, developing, and maintaining PySide2 on Gentoo. It can't have been easy, because nothing worth doing is. You are unsung galactic heroes and the stuff of open-source legend.

It both amuses and saddens me that passing the `-g python` option to Qt 5's standard `rcc` and `uic` binaries has largely obsoleted `pyside2-tools`. This is progress.

I still maintain a [PySide2-based GUI that imports from the `pyside2uic` package installed by `pyside2-tools` at runtime](https://gitlab.com/betse/betsee), which means I still need `pyside2-tools`. So much for progress.

With permission, I'd also like to host @fordfrog's substantially improved `pyside2-tools` ebuild at [my own overlay](https://github.com/leycec/raiagent). One can't have enough redundancy, I always say.

In conclusion, Happy New Year and may 2021 be safer, simpler, and stabler than 2020 for all! :fireworks:
